### PR TITLE
[EYE-70] Refresh user credentials when they expire

### PR DESCRIPTION
**Pull request by tucker-moore**
# Description
This PR adds credential refreshing to the mobile app. I performed a pretty big auth refactor to make this easier. it should also make adding endpoints and calling them from components easier.

# Testing
- `npm test`, some tests had to be removed, and many refactored, but the same tests still fail and the same still pass
- Login flow still works

### CODE REVIEW

File: app/human-detector-app/package.json [hunk 1]
```diff
@@ -38,7 +38,7 @@
     "expo-device": "~4.2.0",
     "expo-font": "~10.1.0",
     "expo-linking": "~3.1.0",
-    "expo-modules-core": "^0.11.8",
+    "expo-modules-core": "~0.9.2",
```
Comment (matthew-latanafrancia) [inline app/human-detector-app/package.json:41]: Why did `expo-modules-core` have to changed here?

File: app/human-detector-app/screens/LoginScreen.tsx [hunk 2]
```diff
@@ -15,10 +16,22 @@ interface Props {
...
   return (
     <View>
-      <KeyCloakButton onTokenResponse={onTokenResponse} />
+      <Button
+        title="Login"
+        onPress={() => {
+          const clientId = Constants.manifest?.extra?.clientId;
+          const issuerOrigin = Constants.manifest?.extra?.keycloakUrl;
+          const issuer = new URL('/realms/users', issuerOrigin);
+          const redirectUri = makeRedirectUri();
```
Comment (matthew-latanafrancia) [inline app/human-detector-app/screens/LoginScreen.tsx:28]: Is fixing redirectUri for the managed workflow gonna be something in a different PR?

File: app/human-detector-app/services/backendService.ts [hunk 3]
```diff
@@ -1,110 +1,115 @@
...
-      headers: {
-        Authorization: BEARER + userAccessToken,
-        Accept: 'application/json',
-      },
-    };
+  /**
+   * This method will send a notification token to the backend to be stored
+   * through our endpoints.  This is done at login.
+   *
+   * @param userIdFromLogin : userId of the user that just logged in
```
Comment (matthew-latanafrancia) [inline app/human-detector-app/services/backendService.ts:62]: Could you delete these extra `@params` from the comment since you got rid of them in the function call?

File: app/human-detector-app/src/auth/tokenManager.ts [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+      );
+    }
+    return this.tokenResponse.accessToken;
+  }
+
+  /**
+   * Get the user from their identity token, assuming there is one.
+   */
+  public getUser(): User {
+    // FIXME: stupid!
```
Comment (matthew-latanafrancia) [inline app/human-detector-app/src/auth/tokenManager.ts:58]: what is this fixme about exactly?

File: app/human-detector-app/src/auth/tokenManager.ts [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    discovery: DiscoveryDocument,
+    clientId: string
+  ) {
+    return new this(response, discovery, clientId);
+  }
+
+  /**
+   * Get the access token, refreshing as needed.
+   */
+  public async getAccessToken(): Promise<string> {
```
Comment (matthew-latanafrancia) [inline app/human-detector-app/src/auth/tokenManager.ts:44]: How exactly did you test that this was working?  When I was trying to work with shouldRefresh() I was having issues with it always refreshing no matter what.

**Review by matthew-latanafrancia — COMMENTED** [review]
have some questions.  thank u for refactoring my code

File: app/human-detector-app/package.json (hunk 1 above)
Comment (tucker-moore) [reply app/human-detector-app/package.json:41]: Figured I would sneak a fix in with this PR, I pulled the wrong version in #39 and Expo was complaining about it

File: app/human-detector-app/screens/LoginScreen.tsx (hunk 2 above)
Comment (tucker-moore) [reply app/human-detector-app/screens/LoginScreen.tsx:28]: Yeah I think I'll add that in a separate PR if that's OK, this one is pretty big as-is

File: app/human-detector-app/src/auth/tokenManager.ts (hunk 4 above)
Comment (tucker-moore) [reply app/human-detector-app/src/auth/tokenManager.ts:58]: It's parsing the user from the ID token every time rather than caching, which honestly probably isn't that big of a deal unless we start attaching more complicated information to the user object.

File: app/human-detector-app/src/auth/tokenManager.ts (hunk 5 above)
Comment (tucker-moore) [reply app/human-detector-app/src/auth/tokenManager.ts:44]: I assumed it wasn't refreshing every time, but I just tested and it looks like it is. I dug a little bit and it looks like AuthSession has a refresh margin margin of 10 minutes by default ([source](https://github.com/expo/expo/blob/fe3bd0dc11905e694d2c3ad492e127135c8f2157/packages/expo-auth-session/src/TokenRequest.ts#L41-L43)), but our tokens are configured to expire after 5 minutes. I could change the code, or we could just bump the lifetime of our tokens up to something like 30 minutes.

This should be the only time we have to refresh; our refresh tokens (currently) don't expire, so we can refresh whenever we want to. The backend service calls `getAccessToken()` before each request, so I figured that would be a good time to refresh our tokens if we need to.

File: app/human-detector-app/src/auth/tokenManager.ts (hunk 4 above)
Comment (matthew-latanafrancia) [reply app/human-detector-app/src/auth/tokenManager.ts:58]: Oh actually that makes sense.  Ill try to draw up my plans for those objects this next sprint so we can get a better idea

File: app/human-detector-app/src/auth/tokenManager.ts (hunk 5 above)
Comment (matthew-latanafrancia) [reply app/human-detector-app/src/auth/tokenManager.ts:44]: Would bumping it up to 30 minutes be against like some standard?  If that means we don't have to refresh every time I think it makes more sense to me.  What do you think?

Comment (tucker-moore) [reply app/human-detector-app/src/auth/tokenManager.ts:44]: There doesn't seem to be a standard, based on [the spec](https://www.rfc-editor.org/rfc/rfc6749#section-4.2.2) I think we're good as long as our tokens expire. I wasn't able to find anything helpful just Googling around either. I think a bump from 5 minutes to 30 minutes is reasonable, otherwise we would be constantly refreshing our tokens during regular usage

Comment (matthew-latanafrancia) [reply app/human-detector-app/src/auth/tokenManager.ts:44]: That's fine by me if you're ok with making that change

Comment (tucker-moore) [reply app/human-detector-app/src/auth/tokenManager.ts:44]: Sounds good. Bumped the expiration time up to 20 minutes and manually tested login with some logging thrown in, it no longer refreshes every time

**Review by matthew-latanafrancia — APPROVED** [review]
changed refresh token time offline. lgtm!
